### PR TITLE
feat(Vectorize): Add a vectorized field to semantic graph FunctionExpr

### PR DESCRIPTION
**Pull request by skartikey**
The field vectorized has been added with type FunctionExpr
The filed is added to Rust, Go and flatbuffers schema
Serialization tets for flatbuffers/Go/Rust is updated

Fixes: https://github.com/influxdata/flux/issues/4183

### CODE REVIEW

**Comment by skartikey** [comment]
**Error -** 

```
thread '<unnamed>' panicked at 'range start index 18446744073709551615 out of range for slice of length 0', flux-core/src/semantic/flatbuffers/mod.rs:1038:28
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
SIGABRT: abort
PC=0x7fff2046992e m=3 sigcode=0

goroutine 0 [idle]:
runtime: unknown pc 0x7fff2046992e
stack: frame={sp:0x700004be31a8, fp:0x0} stack=[0x700004b64338,0x700004be3f38)
0x0000700004be30a8:  0x0000000000000000  0x0000000000000000
0x0000700004be30b8:  0x0000000000000000  0x0000000000000000
0x0000700004be30c8:  0x0000000000000000  0x0000000000000000
0x0000700004be30d8:  0x0000000000000000  0x0000000000000000
0x0000700004be30e8:  0x0000000000000000  0x0000000000000000
0x0000700004be30f8:  0x0000000000000000  0x0000700004be31b8
0x0000700004be3108:  0x0000000004b2bfb8  0x0000000000000010
0x0000700004be3118:  0x0000000000000010  0x0000700004be31a0
0x0000700004be3128:  0x000000000467490a  0x0000000000000000
0x0000700004be3138:  0x00007fff807a6000  0x0000000000000000
0x0000700004be3148:  0x00007fff202d9f3b  0x0000700004be31e8
0x0000700004be3158:  0x0000000004b2be80  0x0000002000000000
0x0000700004be3168:  0x000000000d3cde03  0x0000000000000001
0x0000700004be3178:  0x0000000000000000  0x0000000004aff2e0
0x0000700004be3188:  0x0000000000000001  0x0000000004aa7c98
0x0000700004be3198:  0x0000700004be3470  0x0000700004be3220
0x0000700004be31a8: <0x00007fff204985bd  0x0000000000000003
0x0000700004be31b8:  0x0000000004aff2e0  0x0000700004be4000
0x0000700004be31c8:  0x0000700004be31e8  0x0000700004be3210
0x0000700004be31d8:  0x00007fff203ed406  0x0000000000000004
0x0000700004be31e8:  0x00000000fffff9df  0xffffffff04aa833f
0x0000700004be31f8:  0x0000000000000000  0x0000000004aa7c98
0x0000700004be3208:  0x0000700004be3470  0x0000700004be3220
0x0000700004be3218:  0x000000000461a5f9  0x0000700004be32e0
0x0000700004be3228:  0x00000000045eeb40  0x0000000004b2c258
0x0000700004be3238:  0x0000000000000001  0x0000000000000000
0x0000700004be3248:  0x00000000045edb7e  0x0000700004be3290
0x0000700004be3258:  0x0000000000000001  0x0000000004b2bfb8
0x0000700004be3268:  0x0000000000000002  0x0000000000000000
0x0000700004be3278:  0x0000000004aff2e0  0x0000700004be32a0
0x0000700004be3288:  0x0000000000000001  0x0000700004be32d4
0x0000700004be3298:  0x0000000004667c20  0x0000700004be3230
runtime: unknown pc 0x7fff2046992e
stack: frame={sp:0x700004be31a8, fp:0x0} stack=[0x700004b64338,0x700004be3f38)
0x0000700004be30a8:  0x0000000000000000  0x0000000000000000
0x0000700004be30b8:  0x0000000000000000  0x0000000000000000
0x0000700004be30c8:  0x0000000000000000  0x0000000000000000
0x0000700004be30d8:  0x0000000000000000  0x0000000000000000
0x0000700004be30e8:  0x0000000000000000  0x0000000000000000
0x0000700004be30f8:  0x0000000000000000  0x0000700004be31b8
0x0000700004be3108:  0x0000000004b2bfb8  0x0000000000000010
0x0000700004be3118:  0x0000000000000010  0x0000700004be31a0
0x0000700004be3128:  0x000000000467490a  0x0000000000000000
0x0000700004be3138:  0x00007fff807a6000  0x0000000000000000
0x0000700004be3148:  0x00007fff202d9f3b  0x0000700004be31e8
0x0000700004be3158:  0x0000000004b2be80  0x0000002000000000
0x0000700004be3168:  0x000000000d3cde03  0x0000000000000001
0x0000700004be3178:  0x0000000000000000  0x0000000004aff2e0
0x0000700004be3188:  0x0000000000000001  0x0000000004aa7c98
0x0000700004be3198:  0x0000700004be3470  0x0000700004be3220
0x0000700004be31a8: <0x00007fff204985bd  0x0000000000000003
0x0000700004be31b8:  0x0000000004aff2e0  0x0000700004be4000
0x0000700004be31c8:  0x0000700004be31e8  0x0000700004be3210
0x0000700004be31d8:  0x00007fff203ed406  0x0000000000000004
0x0000700004be31e8:  0x00000000fffff9df  0xffffffff04aa833f
0x0000700004be31f8:  0x0000000000000000  0x0000000004aa7c98
0x0000700004be3208:  0x0000700004be3470  0x0000700004be3220
0x0000700004be3218:  0x000000000461a5f9  0x0000700004be32e0
0x0000700004be3228:  0x00000000045eeb40  0x0000000004b2c258
0x0000700004be3238:  0x0000000000000001  0x0000000000000000
0x0000700004be3248:  0x00000000045edb7e  0x0000700004be3290
0x0000700004be3258:  0x0000000000000001  0x0000000004b2bfb8
0x0000700004be3268:  0x0000000000000002  0x0000000000000000
0x0000700004be3278:  0x0000000004aff2e0  0x0000700004be32a0
0x0000700004be3288:  0x0000000000000001  0x0000700004be32d4
0x0000700004be3298:  0x0000000004667c20  0x0000700004be3230

goroutine 1 [syscall]:
runtime.cgocall(0x43669a0, 0xc000391678)
	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/cgocall.go:156 +0x5c fp=0xc000391650 sp=0xc000391618 pc=0x400787c
github.com/influxdata/flux/libflux/go/libflux._Cfunc_flux_semantic_marshal_fb(0xd3e2760, 0xc0001e4030)
	_cgo_gotypes.go:348 +0x4d fp=0xc000391678 sp=0xc000391650 pc=0x41f9aed
github.com/influxdata/flux/libflux/go/libflux.(*SemanticPkg).MarshalFB.func1(0x46c74a0, 0x46c5100)
	/Users/sunilkartikey/Code/influxdata/flux/libflux/go/libflux/analyze.go:26 +0x74 fp=0xc0003916b8 sp=0xc000391678 pc=0x41fa1f4
github.com/influxdata/flux/libflux/go/libflux.(*SemanticPkg).MarshalFB(0xc000284210)
	/Users/sunilkartikey/Code/influxdata/flux/libflux/go/libflux/analyze.go:26 +0x7a fp=0xc000391770 sp=0xc0003916b8 pc=0x41f9d1a
github.com/influxdata/flux/internal/cmd/builtin/cmd.compilePackage({0x7ffeefbff76c, 0xf}, {0xc00011e6f0, 0x2b}, 0x100000004041811)
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/compile.go:35 +0xe8 fp=0xc000391868 sp=0xc000391770 pc=0x435d708
github.com/influxdata/flux/internal/cmd/builtin/cmd.generate.func1({0xc00011e6f0, 0x2b})
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/generate.go:128 +0x616 fp=0xc000391ae8 sp=0xc000391868 pc=0x435e796
github.com/influxdata/flux/internal/cmd/builtin/cmd.walkDirs({0xc00011e6f0, 0x2b}, 0xc000391d70)
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/generate.go:278 +0x5b fp=0xc000391b60 sp=0xc000391ae8 pc=0x43625bb
github.com/influxdata/flux/internal/cmd/builtin/cmd.walkDirs({0xc000033a58, 0x16}, 0xc000391d70)
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/generate.go:284 +0x135 fp=0xc000391bd8 sp=0xc000391b60 pc=0x4362695
github.com/influxdata/flux/internal/cmd/builtin/cmd.walkDirs({0xc00062dde9, 0x7}, 0xc000391d70)
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/generate.go:284 +0x135 fp=0xc000391c50 sp=0xc000391bd8 pc=0x4362695
github.com/influxdata/flux/internal/cmd/builtin/cmd.walkDirs({0x47292df, 0x1}, 0xc000391d70)
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/generate.go:284 +0x135 fp=0xc000391cc8 sp=0xc000391c50 pc=0x4362695
github.com/influxdata/flux/internal/cmd/builtin/cmd.generate(0x4b690c0, {0x4729898, 0x6, 0x6})
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/generate.go:57 +0xe5 fp=0xc000391db0 sp=0xc000391cc8 pc=0x435de85
github.com/spf13/cobra.(*Command).execute(0x4b690c0, {0xc0000c82a0, 0x6, 0x6})
	/Users/sunilkartikey/Code/influxdata/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:762 +0x60e fp=0xc000391e70 sp=0xc000391db0 pc=0x435770e
github.com/spf13/cobra.(*Command).ExecuteC(0x4b69320)
	/Users/sunilkartikey/Code/influxdata/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852 +0x2dc fp=0xc000391f28 sp=0xc000391e70 pc=0x4357c1c
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/sunilkartikey/Code/influxdata/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
github.com/influxdata/flux/internal/cmd/builtin/cmd.Execute()
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/cmd/root.go:18 +0x25 fp=0xc000391f70 sp=0xc000391f28 pc=0x4365d05
main.main()
	/Users/sunilkartikey/Code/influxdata/flux/internal/cmd/builtin/main.go:6 +0x17 fp=0xc000391f80 sp=0xc000391f70 pc=0x43667b7
runtime.main()
	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/proc.go:255 +0x227 fp=0xc000391fe0 sp=0xc000391f80 pc=0x40398e7
runtime.goexit()
	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc000391fe8 sp=0xc000391fe0 pc=0x4066781

rax    0x0
rbx    0x700004be4000
rcx    0x700004be31a8
rdx    0x0
rdi    0xb03
rsi    0x6
rbp    0x700004be31d0
rsp    0x700004be31a8
r8     0x700004be305f
r9     0x0
r10    0x0
r11    0x246
r12    0xb03
r13    0x1
r14    0x6
r15    0x16
rip    0x7fff2046992e
rflags 0x246
cs     0x7
fs     0x0
gs     0x0
exit status 2
stdlib/gen.go:4: running "go": exit status 1
make: *** [stdlib/packages.go] Error 1
```

File: semantic/flatbuffers.go [hunk 1]
```diff
@@ -570,6 +570,16 @@ func (e *FunctionExpression) FromBuf(fb *fbsemantic.FunctionExpression) error {
 		return err
 	}
 
+	fbVectorized := fb.Vectorized(nil)
+	if fbVectorized == nil {
+		return errors.New(codes.Internal, "missing function vectorized")
```
Comment (wolffcm) [inline semantic/flatbuffers.go:575]: I would expect this function to be optional. so it should not be an error if it's not there.

File: libflux/flux-core/src/semantic/flatbuffers/mod.rs [hunk 2]
```diff
@@ -1104,6 +1106,71 @@ impl<'a, 'b> SerializingVisitorState<'a, 'b> {
             },
         ))
     }
+
+    fn create_function_expression(&mut self, func: &semantic::nodes::FunctionExpr) -> WIPOffset<fbsemantic::FunctionExpression<'a>> {
```
Comment (wolffcm) [inline libflux/flux-core/src/semantic/flatbuffers/mod.rs:1110]: Is this code a copy from above? I think it should be possible to do this without duplicating code. 

I think what needs to happen is that the visitor that walks over the semantic graph needs to be updated to visit the `vectorized` field of a function expression. If you do that then I would expect that the expression would be on the `expr_stack`.

**Review by wolffcm — COMMENTED** [review]
I'm not sure about the panic that's happening. Is there a particular Go test that is triggering it? I think if you could reproduce it in a Rust unit test you might get a better idea of what's going on. Probably a Rust unit test would just need to attempt to serialize a function that has the `vectorized` field set?

File: libflux/flux-core/src/semantic/flatbuffers/tests.rs [hunk 3]
```diff
@@ -495,6 +495,9 @@ fn compare_exprs(
                 }
                 block_len += 1;
             }
+            if let Some(vectorized) = &semantic_fe.vectorized {
+                compare_exprs(&vectorized, fb_fe.vectorized_type(), &fb_fe.vectorized())?;
+            }
```
Comment (wolffcm) [inline libflux/flux-core/src/semantic/flatbuffers/tests.rs:500]: The test case that uses this code generates semantic graph by parsing and analyzing raw Flux text. That won't populate the `vectorized` field unfortunately. Can you somehow make it so we actually test the new code you wrote? I see two ways to do it:
- modify a function expression in the test case before it's serialized such that the `vectorized` field has something in it.
- Add a special test case with a semantic graph function expression created by hand that contains a function expression with a populated `vectorized` field.

File: semantic/flatbuffers_test.go [hunk 4]
```diff
@@ -292,6 +292,8 @@ func getFnExprFlatBuffer() (*semantic.Package, []byte) {
 	fbsemantic.FunctionExpressionAddLoc(b, exprLoc)
 	fbsemantic.FunctionExpressionAddTyp(b, funTy)
 	fbsemantic.FunctionExpressionAddTypType(b, fbsemantic.MonoTypeFun)
+	fbsemantic.FunctionExpressionAddVectorized(b, 0)
+	fbsemantic.FunctionExpressionAddVectorizedType(b, fbsemantic.ExpressionFunctionExpression)
```
Comment (wolffcm) [inline semantic/flatbuffers_test.go:296]: So to properly test your code, you'll need to update the massive set of nested structures below (The `want := ...` statement that describes the desired semantic graph) to include an actual function expression in the `vectorized` field.  

In addition there needs to be an actual vectorized function expression in the flatbuffer input to deserialization. It doesn't matter what the function expression is, just that it's there and you can show that it's properly deserialized.

File: internal/fbsemantic/semantic.fbs [hunk 5]
```diff
@@ -276,6 +276,7 @@ table FunctionExpression {
   params:[FunctionParameter];
   body:Block;
   typ:MonoType;
+  vectorized:Expression;
```
Comment (wolffcm) [inline internal/fbsemantic/semantic.fbs:279]: Can this be a `FunctionExpression`?

File: semantic/graph.go [hunk 6]
```diff
@@ -572,7 +572,8 @@ type FunctionExpression struct {
 	Defaults   *ObjectExpression
 	Block      *Block
 
-	typ MonoType
+	typ        MonoType
+	Vectorized Expression
```
Comment (wolffcm) [inline semantic/graph.go:576]: This could be made into a function expression, yes?

**Review by wolffcm — CHANGES_REQUESTED** [review]
This looks pretty close, but I had a question about the type of the field and some suggestions on how to improve the tests.

File: libflux/flux-core/src/semantic/nodes.rs [hunk 7]
```diff
@@ -981,7 +983,7 @@ pub struct FunctionExpr {
     pub params: Vec<FunctionParameter>,
     pub body: Block,
 
-    pub vectorized: Option<Box<FunctionExpr>>,
+    pub vectorized: Option<Expression>,
```
[the remark this answers is not in the document]
Comment (skartikey) [reply libflux/flux-core/src/semantic/nodes.rs:986]: As there was less refactoring required when using Expression, I have used it with calling FunctionExpr variant at required places.
I have changed it now to FunctionExpr and did require refactoring.

File: libflux/flux-core/src/semantic/flatbuffers/tests.rs [hunk 8]
```diff
@@ -108,6 +116,79 @@ re !~ /foo/
     }
 }
 
+#[test]
+fn test_serialize_vectorization() {
```
Comment (wolffcm) [inline libflux/flux-core/src/semantic/flatbuffers/tests.rs:120]: Can you add a comment here that describes why you needed to add this extra test case? In other words, explain that the normal parse/analyze sequence won't populate the `vectorized` field, so you need to do somethign special here to do that.

File: semantic/flatbuffers_test.go [hunk 9]
```diff
@@ -196,7 +201,114 @@ func getUnaryOpFlatBuffer() (*semantic.Package, []byte) {
 func getFnExprFlatBuffer() (*semantic.Package, []byte) {
```
Comment (wolffcm) [inline semantic/flatbuffers_test.go:201]: You could refactor this function with the one below it to get rid of some duplicate code that constructs the assignment statement in the flatbuffer, as well as the Go structure. But I'll leave that up to you.

File: semantic/flatbuffers_test.go [hunk 10]
```diff
@@ -292,178 +404,142 @@ func getFnExprFlatBuffer() (*semantic.Package, []byte) {
...
-	pt := getFnPolyType(b)
-	asnLoc := getFBLoc(b, "1:1", "1:36", src)
-	fbsemantic.NativeVariableAssignmentStart(b)
-	fbsemantic.NativeVariableAssignmentAddLoc(b, asnLoc)
-	fbsemantic.NativeVariableAssignmentAddTyp(b, pt)
-	fbsemantic.NativeVariableAssignmentAddIdentifier(b, id)
-	fbsemantic.NativeVariableAssignmentAddInit_(b, fe)
-	fbsemantic.NativeVariableAssignmentAddInit_type(b, fbsemantic.ExpressionFunctionExpression)
-	nva := fbsemantic.NativeVariableAssignmentEnd(b)
+func funcExpression(fe *semantic.FunctionExpression) *semantic.FunctionExpression {
```
Comment (wolffcm) [inline semantic/flatbuffers_test.go:415]: This function could do with a comment that describes what the argument is for, as it won't be obvious to someone reading the code.

**Review by wolffcm — APPROVED** [review]
This looks good to me. I had some feedback about comments and such but I think this is fine to merge.

File: libflux/flux-core/src/semantic/flatbuffers/tests.rs (hunk 3 above)
Comment (skartikey) [reply libflux/flux-core/src/semantic/flatbuffers/tests.rs]: This has been implemented.

File: semantic/flatbuffers_test.go (hunk 4 above)
Comment (skartikey) [reply semantic/flatbuffers_test.go]: Thanks for helping out in writing this test case. Done!

File: semantic/flatbuffers_test.go (hunk 10 above)
Comment (skartikey) [reply semantic/flatbuffers_test.go]: comments added

File: semantic/flatbuffers_test.go (hunk 9 above)
Comment (skartikey) [reply semantic/flatbuffers_test.go:201]: Done!

**Pull request merged**